staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -723,22 +738,21 @@ func (qs *queueSet) selectQueueLocked() *queue {
 	// for the next round.  This way the non-selected queues
 	// win in the case that the virtual finish times are the same
 	qs.robinIndex = minIndex
-	// according to the original FQ formula:
-	//
-	//   Si = MAX(R(t), Fi-1)
-	//
-	// the virtual start (excluding the estimated cost) of the chose
-	// queue should always be greater or equal to the global virtual
-	// time.
-	//
-	// hence we're refreshing the per-queue virtual time for the chosen
-	// queue here. if the last start R (excluded estimated cost)
-	// falls behind the global virtual time, we update the latest virtual
-	// start by: <latest global virtual time> + <previously estimated cost>
-	previouslyEstimatedServiceTime := float64(minQueue.seatsInUse) * qs.estimatedServiceSeconds
-	if qs.virtualTime > minQueue.virtualStart-previouslyEstimatedServiceTime {
+
+	// Next comes the anti-windup hack.
+	// We need a hack because all non-empty queues are allocated the same
+	// number of seats.  A queue that can not use all those seats and does
+	// not go empty accumulates a progresively earlier `virtualStart` compared
+	// to queues that are using more than they are allocated.
+	// The following hack addresses the first side of that inequity,
+	// by insisting that dispatch in the virtual world not precede arrival.
+	var virtualStartBound = oldestReqFromMinQueue.arrivalR
+	if minQueue.virtualStart < virtualStartBound {
```

Comment:
Thinking about this - I can imagine that after adjusting the `minQueue` is no longer the queue that we should choose.

Shouldn't we actually do those adjustments above (when choosing the queue in line 701-719)?

Reply:
Better yet, modify the code that currently introduces the suppressed situation.